### ISSUE TITLE
Redaction modifies original object when redacted property is absent

### ISSUE THREAD

**Issue by dougmoscrop** [pinojs/pino#521]
`pino-noir` does not exhibit this behaviour; it was quite a surprise that my logged object was being manipulated. Made some `joi` validation very unhappy, but only when the log level happened to be DEBUG ;)

```js
const pino = require('.');
const log = pino({ redact:  { paths: ['obj.foo'] } });

const first = {}; <--- obj is empty
const second = { foo: 'test' };

log.info({ obj: first }, 'first');
console.log('first', first); <----- obj now has property set to null

log.info({ obj: second }, 'second');
console.log('second', second);
```

Outputs:

```
{"msg":"first","obj":{}}
first { foo: null } 
```
```
{"msg":"second","obj":{"foo":"[Redacted]"}}
second { foo: 'test' }
```

**Comment by mcollina** [comment]
Thanks for reporting. This looks like a bug to me. I’m not sure if here or in our fast-redact dependency.

Would you like to send a PR? This is an extremely busy period for all of us.

**Comment by dougmoscrop** [comment]
I will try!

**Comment by jsumners** [comment]
@dougmoscrop any luck figuring out where the bug is?

**Comment by dougmoscrop** [comment]
Sorry, no, been super busy.

**Comment by jsumners** [comment]
This is definitely a bug. It should not be closed until we resolve it.

**Comment by dougmoscrop** [comment]
Whoops that was an accident

**Comment by jsumners** [comment]
The trigger is the nested object:

```js
// This works correctly
const log = pino({ redact: ['foo'] })
const obj = {}
log.info(obj)
```

// This triggers the bug
const log = pino({ redact: ['foo.bar'] })
const obj = {}
log.info({foo: obj})
```

**Comment by jsumners** [comment]
During the `stringify` process (https://github.com/pinojs/pino/blob/4b5db2756c6a5be60d459505c2643775a4a2179f/lib/tools.js#L126) the following code gets generated by https://github.com/davidmarkclements/fast-redact/blob/3bc8dbe3547542b22b44262067d7390d5efc37b8/lib/restorer.js#L16 :

```js
(function(o
/*``*/) {

    const secret = this.secret
    
      try { o.bar = secret["bar"].val } catch (e) {}
      secret["bar"].val = null
    
    
    return o
  
})
```

Notice that this is modifying the original object passed in to the anonymous function. This is the point at which the source object is being modified.

@davidmarkclements do you have time to figure this out? I'm not following your magic well enough to even figure out where the code comes from in that anonymous function.

I have started a PR for the resolution of this issue. Feel free to update it -- https://github.com/pinojs/pino/pull/524

**Comment by davidmarkclements** [comment]
I don't think this is the culprit - from memory `secret` isn't the object itself but a piece of meta data for the object - all this particular null is doing is clearing the meta data because it's just been used 

This will require more digging - my time is extremely limited right now - I can advise from afar but I can't currently fix for a while

**Comment by jsumners** [comment]
`o` is the object, not `secret`. And `o.bar` is getting assigned `null` because `secret["bar"]` doesn't exist. Right?

**Comment by davidmarkclements** [comment]
Only if this function is called more than once - the secret val is set to null no matter what (notice it's outside the catch) - so if it was called a second time it could then set o.bar to null - unless secret val is initialised to null in the first place then it only needs to be called once. 

At any rate I wonder if adding an `if ('bar' in o)` check before the assignment would solve it - preferably without causing a performance hit

**Comment by davidmarkclements** [comment]
Of course it's not that simple because it's precompiled (see resetTmpl function in same file, it's the reset variable) and the path can be any level of nesting

**Comment by jsumners** [comment]
See https://youtu.be/iakSMejb2Z0 . It clearly shows that this is happening on the first call.

I am trying to create a failing test in `fast-redact` but not having any luck. But I'm out of time for now. Need to go get a 🚴 in before a Naughton twins concert this afternoon.

**Comment by jsumners** [comment]
Failing test added in https://github.com/davidmarkclements/fast-redact/pull/6. I had a typo last weekend (`path` instead of `paths`).

**Comment by mcollina** [comment]
This should be fixed in fast-redact 1.3.0. Can you please confirm everything now works correctly @dougmoscrop?

**Comment by dougmoscrop** [comment]
Yes it looks good to me!
